utils/virtual_machine/utils.py

```diff
@@ -0,0 +1,40 @@
+import pytest
+from utils import context
+
+
+def parametrize_virtual_machines(bugs: list[dict] = None):
+    """ You can set multiple bugs for a single test case. 
+    If you want to set a bug for a specific VM, you can set the vm_name or vm_cpu or weblog_variant in the bug dictionary (using one or more fields). 
+    ie: 
+    - Marks as bug for vm with name "vm1" and weblog_variant "app1"
+    *     @parametrize_virtual_machines(bugs=[{"vm_name":"vm1", "weblog_variant":"app1", "reason": "APMON-1576"}])
+    - Marks as bug for vm with cpu type "amd64" and weblog_variant "app1"
+    *     @parametrize_virtual_machines(bugs=[{"vm_cpu":"amd64", "weblog_variant":"app1", "reason": "APMON-1576"}])
+    - Marks as bug for all vm with cpu type "amd64" and all weblogs
+    *     @parametrize_virtual_machines(bugs=[{"vm_cpu":"amd64","reason": "APMON-1576"}])
+    Reason is mandatory for each bug and it MUST reference to a JIRA ticket.
+    """
+
+    def decorator(func):
+
+        parameters = []
+
+        for vm in getattr(context.scenario, "required_vms", []):
+            bug_found = False
+            if bugs and hasattr(bugs, "__len__") and len(bugs) > 0:
+                for bug in bugs:
+                    if (
+                        ("vm_name" not in bug or vm.name == bug["vm_name"])
+                        and (not "vm_cpu" in bug or vm.os_cpu == bug["vm_cpu"])
+                        and (not "weblog_variant" in bug or context.weblog_variant == bug["weblog_variant"])
+                    ):
```

Comment:
Maybe I'm daft, but isn't the predicate almost-but-not-exactly inverted here? I would expect that the condition for marking the test as known failure would be:
```suggestion
                        ("vm_name" in bug and vm.name == bug["vm_name"])
                        or ("vm_cpu" in bug and vm.os_cpu == bug["vm_cpu"])
                        or ("weblog_variant" in bug and context.weblog_variant == bug["weblog_variant"])
                    ):
```
I'm sure it's just me missing something here.

Reply:
I think it's ok. But please, confirm me this after this explanation:

for example:
* mark as bug all machines with cpu==amd64 AND weblog_variant=="xyz"
`bugs=[{"vm_cpu": "amd64", "weblog_variant":"xyz", "reason": "APMON-1576"}]`

* mark as bug all machines with vm_name="vm1" AND cpu==amd64 AND weblog_variant=="xyz"
`bugs=[{"vm_name": "vm1", "vm_cpu": "amd64", "weblog_variant":"xyz", "reason": "APMON-1576"}]`
I mean, if I have a machine "vm1" with cpu "arm64" is going to execute the test

Reply:
Oh, got it – if there _is_ a bug marker, then the vm needs to match all specified fields to be a known failing one, and it of course implicitly matches any unspecified fields. Thanks for bearing with me!